### DISCUSSION TITLE
A question about the code in the first tutorial (Ising model)

### DISCUSSION THREAD

**Discussion by Dzhou12** [netket/netket#1472] [Q&A]
I'm trying to learn this toolbox through its tutorial(transverse field Ising model), and after reading its 6th part(Neural network quantum state), when I add a snippet and tried to write  
x=jnp.array([1,2,3])
dense = nn.Dense(features=x.shape[0])
y=dense(x)
to see how the input configuration will be manipulated though the network, there is an error on the 3rd line:
CallCompactUnboundModuleError: Can't call compact methods on unbound modules
I can't understand this error, how is it generated?

**Comment by MarcMachaczek** [comment]
This has something to do with the nature of jax and flax, not primarily netket. Following a functional programming paradigm, parameters of the flax model are not stored as attributes of the nn.module (in this case nn.Dense). To apply a forward pass of nn.Dense use
`y = dense.apply(parameters, x)`
after you initialized the model with something like 
`params = dense.initialize(jax.random.PRNG(0), x)`
Notice that to apply the model, you need to explicitly provide its parameters. Think of a flax module as just being a function template which has knowledge of how it can calculate an output given some input and its own variables (like parameters, model state etc.) This is necessary to properly make use of jax transformations like jit, which require pure functions that, for instance, cannot manipulate the state of objects outside the function. 
I strongly recommend working and reading through https://flax.readthedocs.io/en/latest/guides/flax_basics.html, as well as https://flax.readthedocs.io/en/latest/developer_notes/module_lifecycle.html for more in-depth information about why flax modules work like they do.
In short, these "empty function templates" are unbound modules in flax language because their parameters etc are not bound to them. Bound versions are created within the apply call, enabling the __call__ method to access model parameters and perform the forward pass.
